### PR TITLE
Replace services/graph.ts with graph-service

### LINKED ISSUE #35: Replace services/graph.ts with Graph Service

### PR DESCRIPTION
**Pull request by mattheworris**
# Purpose
The goal of this PR is integrate the graph service with the social app template.

Closes #35 

## Solution

- [x] Replace any polkadot.js calls with Graph-Service endpoints.

Started by/Contributions from:
@claireolmstead @JoeCap08055 @saraswatpuneet 

### Change summary
- [x] Added configuration items needed for `graph-service`
- [x] Refactored `GraphService` class
- [x] Updated `docker-compose.yaml` to correctly launch `graph-service-api` and `graph-service-worker`
- [x] Migrated all `graph/dsnpId` endpoints to `graph/msaId`
- [x] Implement scaffolding for graph-service webhook 
      (need to address in follow-on issue #51 )

## Steps to Verify
1. Setup the containers with `docker compose up -d`
2. Sign up/sign in with SIWF
3. Create a post
4. Logout with current user
5. Sign up/sign in with new user account.
6. Navigate to the original user's post.
7. Verify that the follow button is available and executes correctly (there is a delay after the follow request)

## Follow Up

- [ ] #51
- [ ] #52
- [ ] #53

### CODE REVIEW

File: backend/controllers/GraphController.ts [hunk 1]
```diff
@@ -4,30 +4,38 @@ import { HttpStatusCode } from 'axios';
...
       return res.status(HttpStatusCode.NotFound).send();
     }
 
     try {
-      const follows = await GraphService.instance().then((service) => service.getPublicFollows(msaId));
+      const follows = await GraphService.getInstance().then((service) => service.getPublicFollows(msaId));
       return res.status(HttpStatusCode.Ok).send(follows);
     } catch (err) {
-      console.error('Error getting user follows', err);
+      logger.error(`Error getting user follows: ${err}`);
```
Comment (JoeCap08055) [inline backend/controllers/GraphController.ts:38]: Note, this is the canonical way to do error logging with the Pino logger we're using. Also, generally the format to use is `logger.<method>(objectToLog, logMessage)`. Note for error object, the property name of the error object must be `err` (configurable)
```suggestion
      logger.error({ err }, 'Error getting user follows');
```

File: backend/controllers/GraphController.ts [hunk 2]
```diff
@@ -43,16 +51,16 @@ export class GraphController extends BaseController {
...
       return res.status(HttpStatusCode.BadRequest).send();
     }
 
     try {
-      await GraphService.instance().then((service) => service.follow(msaId, parseInt(msaToFollow)));
+      await GraphService.getInstance().then((service) => service.postFollow(msaId, parseInt(msaToFollow)));
       return res.status(HttpStatusCode.Created).send();
     } catch (err: any) {
-      console.error('Error changing user graph: follow', err);
+      logger.error(`Error changing user graph: follow:(${err})`);
```
Comment (JoeCap08055) [inline backend/controllers/GraphController.ts:63]:
```suggestion
      logger.error({ err }, 'Error changing user graph: follow');
```

File: backend/controllers/GraphController.ts [hunk 3]
```diff
@@ -67,17 +75,17 @@ export class GraphController extends BaseController {
...
     }
 
     try {
-      await GraphService.instance().then((service) => service.unfollow(msaId, parseInt(msaToUnfollow)));
+      await GraphService.getInstance().then((service) => service.postUnfollow(msaId, parseInt(msaToUnfollow)));
 
       return res.status(HttpStatusCode.Created).send();
     } catch (err: any) {
-      console.error('Error changing user graph: unfollow', err);
+      logger.error(`Error changing user graph: unfollow ${err}`);
```
Comment (JoeCap08055) [inline backend/controllers/GraphController.ts:88]:
```suggestion
      logger.error({ err }, 'Error changing user graph: unfollow');
```

File: backend/controllers/WebhookController.ts [hunk 4]
```diff
@@ -35,4 +37,17 @@ export class WebhookController extends BaseController {
       }
     }
   }
+  public async graphServiceWebhook(req: Request, res: Response) {
+    try {
+      const response = await GraphServiceWebhook.getInstance().then((service) => service.graphServiceWebhook(req.body));
+      res.send(response).end();
+    } catch (e) {
+      logger.error(`Error handling graph service webhook: ${e}`);
```
Comment (JoeCap08055) [inline backend/controllers/WebhookController.ts:45]:
```suggestion
      logger.error({ err: e }, 'Error handling graph service webhook');
```

File: backend/environment/env.social-app-backend.docker.template [hunk 5]
```diff
@@ -13,6 +13,9 @@ ACCOUNT_SERVICE_URL=http://account-service-api:3000
 # Base URL for the content-publishing-service
 CONTENT_PUBLISHER_URL=http://content-publishing-service-api:3000
 
+# Base URL for the graph-service
+GRAPH_SERVICE_URL=http://0.0.0.0:3002
```
Comment (JoeCap08055) [inline backend/environment/env.social-app-backend.docker.template:17]:
```suggestion
GRAPH_SERVICE_URL=http://graph-service-api:3000
```

File: backend/services/GraphService.ts [hunk 6]
```diff
@@ -50,125 +86,79 @@ export class GraphService {
...
-    });
-
+    const graphsQueryParamsDto: GraphsQueryParamsDto = {
+      dsnpIds: [msaId],
+      privacyType: 'public',
+      graphKeyPairs: [],
+    };
+    logger.debug(
+      `GraphService: getPublicFollows: msaId(${msaId}), graphsQueryParamsDto:(${JSON.stringify(graphsQueryParamsDto)}`
+    );
```
Comment (JoeCap08055) [inline backend/services/GraphService.ts:102]:
```suggestion
    logger.debug({ msaId, graphsQueryParamsDto }, 'GraphService: getPublicFollows');
```

File: backend/services/GraphService.ts [hunk 7]
```diff
@@ -50,125 +86,79 @@ export class GraphService {
...
+    const resp = await this.client.ApiController_getGraphs(null, graphsQueryParamsDto);
+    const userGraphDto: UserGraphDto[] = resp.data;
+    logger.warn(`GraphService: getPublicFollows userGraphDto:(${JSON.stringify(userGraphDto)})`);
```
Comment (JoeCap08055) [inline backend/services/GraphService.ts:105]:
```suggestion
    logger.warn(userGraphDto, 'GraphService: getPublicFollows');
```

File: backend/services/GraphService.ts [hunk 8]
```diff
@@ -50,125 +86,79 @@ export class GraphService {
...
+    const followList: string[] = userGraphDto
+      .map((userGraph) => userGraph.dsnpGraphEdges?.map((edge) => edge.userId.toString()))
+      .filter((item): item is string[] => item !== undefined)
+      .flat();
+    logger.debug(`GraphService: getPublicFollows followList:(${followList})`);
```
Comment (JoeCap08055) [inline backend/services/GraphService.ts:110]:
```suggestion
    logger.debug(followList, 'GraphService: getPublicFollows');
```

File: backend/services/GraphService.ts [hunk 9]
```diff
@@ -50,125 +86,79 @@ export class GraphService {
...
+            connectionType: 'follow',
+          },
+        ],
+      },
+    };
+    const resp = await this.client.ApiController_updateGraph(null, providerGraphDto);
+    // Here we get the reference Id from the BullMQ worker queue
+    // We can setup a webhook to listen for the response when the block with this txn is finalized
+    // REMOVE: For now we will just assume that the transaction is successful in about 14 seconds
+    logger.debug(`REMOVE: DEBUG: Follow Response: resp:(${JSON.stringify(resp.data)})`);
```
Comment (JoeCap08055) [inline backend/services/GraphService.ts:140]:
```suggestion
    logger.debug({ response: response.data }, 'REMOVE: DEBUG: Follow Response');
```

File: backend/services/GraphService.ts [hunk 10]
```diff
@@ -50,125 +86,79 @@ export class GraphService {
...
-      .payWithCapacity(tx)
-      .signAndSend(getProviderKey(), { nonce: await getNonce() }, ({ status, dispatchError }) => {
-        if (dispatchError) {
-          console.error('Graph ERROR: ', dispatchError.toHuman());
-        } else if (status.isInBlock || status.isFinalized) {
-          console.log('Graph Updated: ', status.toHuman());
-        }
-      });
+  public async postUnfollow(actorId: string, objectId: number): Promise<void> {
+    logger.debug(`Unfollow Request: actorId:(${actorId}), objectId:(${objectId})`);
```
Comment (JoeCap08055) [inline backend/services/GraphService.ts:144]:
```suggestion
    logger.debug({ actorId, objectId }, 'Unfollow Request');
```

File: backend/services/GraphService.ts [hunk 11]
```diff
@@ -50,125 +86,79 @@ export class GraphService {
...
+            connectionType: 'follow',
+          },
+        ],
+      },
+    };
+    const resp = await this.client.ApiController_updateGraph(null, providerGraphDto);
+    // Here we get the reference Id from the BullMQ worker queue
+    // We can setup a webhook to listen for the response when the block with this txn is finalized
+    // REMOVE: For now we will just assume that the transaction is successful in about 14 seconds
+    logger.debug(`REMOVE: DEBUG: Unfollow Response: resp:( ${JSON.stringify(resp.data)} )`);
```
Comment (JoeCap08055) [inline backend/services/GraphService.ts:162]:
```suggestion
    logger.debug({ response: resp.data }, 'REMOVE: DEBUG: Unfollow Response');
```

File: backend/services/announce.ts [hunk 12]
```diff
@@ -85,16 +87,16 @@ export const publish = async <T extends BroadcastAnnouncement | ReplyAnnouncemen
...
   // Do NOT wait for all the callbacks. Assume for now that it will work...
-  await api.tx.frequencyTxPayment
+  await chainApi.tx.frequencyTxPayment
     .payWithCapacity(tx)
     .signAndSend(getProviderKey(), { nonce: await getNonce() }, ({ status, dispatchError }) => {
       if (dispatchError) {
         console.error('ERROR: ', dispatchError.toHuman());
       } else if (status.isInBlock || status.isFinalized) {
-        console.log('Message Posted', status.toHuman());
+        logger.info('Message Posted', status.toHuman());
```
Comment (JoeCap08055) [inline backend/services/announce.ts:99]:
```suggestion
        logger.info({ status: status.toJson() }, 'Message Posted');
```

File: backend/controllers/AssetsController.ts [hunk 13]
```diff
@@ -20,7 +20,7 @@ export class AssestsController extends BaseController {
   public async postAssets(req: Request, res: Response) {
     const { files } = req;
     if (!Array.isArray(files) || !files.length) {
-      return res.status(HttpStatusCode.BadRequest).send('Empty asset upload request');
+      return res.status(HttpStatusCode.Accepted).send('No files were uploaded.');
```
Comment (claireolmstead) [inline backend/controllers/AssetsController.ts:23]: I prefer this logic as well.

File: backend/controllers/ProfilesController.ts [hunk 14]
```diff
@@ -1,20 +1,33 @@
...
-    this.router.get('/:dsnpId', this.getProfile.bind(this));
-    this.router.put('/:dsnpId', this.putProfile.bind(this));
+    this.router.get('/:msaId', this.getProfile.bind(this));
+    this.router.put('/:msaId', this.putProfile.bind(this));
   }
 
-  public getProfile(_req: Request, res: Response) {
-    return res.status(HttpStatusCode.Ok).send();
+  public async getProfile(_req: Request, res: Response) {
+    const Profile: Partial<Profile> = {
```
Comment (claireolmstead) [inline backend/controllers/ProfilesController.ts:20]: nit: kinda confusing to have a const and type of the same name. I thought our convention was that only types were uppercase and values were lower case.

**Review by JoeCap08055 — APPROVED** [review]

File: backend/controllers/WebhookController.ts [hunk 15]
```diff
@@ -26,12 +28,25 @@ export class WebhookController extends BaseController {
         service.accountServiceWebhook(req.body)
       );
       res.send(response).end();
-    } catch (e) {
-      logger.error(`Error handling account service webhook: ${e}`);
-      if (e instanceof HttpError) {
-        res.status(e.code).send(e.message).end();
+    } catch (err) {
+      logger.error({ err }, 'Error handling account service webhook:');
```
Comment (claireolmstead) [inline backend/controllers/WebhookController.ts:32]:
```suggestion
      logger.error({ err }, 'Error handling account service webhook.');
```

File: backend/services/GraphService.ts [hunk 16]
```diff
@@ -1,174 +1,118 @@
...
-      const data = zlib.inflateSync(container.compressedPublicGraph);
-      const graphEdges = publicFollowsAvro.fromBuffer(data);
-      return graphEdges;
-    } catch (e) {
-      console.log('Error parsing page', e);
-      return [];
+  private async connect() {
+    if (this._client === undefined) {
+      const api = new OpenAPIClientAxios({
+        definition: openapiJson as unknown as Document,
```
Comment (claireolmstead) [inline backend/services/GraphService.ts:28]: I wish we could do this type better. Going to mess around with it.

File: backend/services/GraphService.ts [hunk 17]
```diff
@@ -1,174 +1,118 @@
...
+    logger.debug({ actorId, objectId }, 'Follow Request');
+    const providerGraphDto: ProviderGraphDto = {
+      dsnpId: actorId,
+      connections: {
+        data: [
+          {
+            dsnpId: objectId.toString(),
+            privacyType: 'public',
+            direction: 'connectionTo',
+            connectionType: 'follow',
```
Comment (claireolmstead) [inline backend/services/GraphService.ts:85]: Is there a reason that we don't want to have these strings be enums or held somewhere more global to be reused?

File: backend/services/GraphWebhookService.ts [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+type GraphServiceWebhookResponse = {
+  referenceId: string;
+  msaId: string;
+  update: string;
+};
+
+type GraphChangeNotification = {
+  msaId: string;
+  update: string;
+};
```
Comment (claireolmstead) [inline backend/services/GraphWebhookService.ts:18]: is there any way we can import these types directly from graph service?

Comment (mattheworris) [reply backend/services/GraphWebhookService.ts:18]: Possibly, this code is a placeholder for now. Some re-work is required in follow-on.

File: backend/services/GraphService.ts (hunk 17 above)
Comment (JoeCap08055) [reply backend/services/GraphService.ts:85]: Really, these should be defined by `graph-service` and exported/imported via the OpenAPI document. I'll look into it; we do something similar with `content-watcher`.

**Pull request merged**